sct-commons/src/main/java/org/lfenergy/compas/sct/commons/DoTypeService.java

```diff
@@ -25,4 +32,59 @@ public Optional<TDOType> findDoType(TDataTypeTemplates tDataTypeTemplates, Predi
+        List<DataAttributeRef> result = new ArrayList<>();
+        // DA -> BDA -> BDA..
+        sdoOrDAService.getSDOOrDAs(tdoType, TDA.class).forEach(tda -> {
+            DataAttributeRef newDataObjectRef = DataAttributeRef.copyFrom(dataRef);
+            newDataObjectRef.getDaName().setName(tda.getName());
+            if(tda.isSetFc()) newDataObjectRef.getDaName().setFc(tda.getFc());
+
+            // STRUCT type (BType=STRUCT) refer to BDA, otherwise it is DA
+            if(tda.isSetBType() && tda.getBType().equals(TPredefinedBasicTypeEnum.STRUCT)) {
+                daTypeService.findDaType(dtt, tdaType -> tda.isSetType() && tdaType.getId().equals(tda.getType()))
+                        .ifPresent(nextDaType -> result.addAll(getDataAttributesFromBDA(dtt, nextDaType, newDataObjectRef)));
+            } else {
+                updateFromAbstractDataAttribute(tda, newDataObjectRef.getDaName());
+                result.add(newDataObjectRef);
+            }
+        });
+        // SDO -> SDO -> SDO..
+        sdoOrDAService.getSDOOrDAs(tdoType, TSDO.class)
+                .forEach(tsdo -> findDoType(dtt, tdoType1 -> tsdo.isSetType() && tdoType1.getId().equals(tsdo.getType()))
+                        .ifPresent(nextDoType -> {
+                            DataAttributeRef newDataAttributeRef = DataAttributeRef.copyFrom(dataRef);
+                            newDataAttributeRef.getDoName().getStructNames().add(tsdo.getName());
+                            if(nextDoType.isSetCdc()) newDataAttributeRef.getDoName().setCdc(nextDoType.getCdc());
+
+                            result.addAll(getAllSDOAndDA(dtt, nextDoType, newDataAttributeRef));
+                        }));
+        return result;
+    }
+
+    private List<DataAttributeRef> getDataAttributesFromBDA(TDataTypeTemplates dtt, TDAType tdaType1, DataAttributeRef dataAttributeRef) {
+        List<DataAttributeRef> result = new ArrayList<>();
+        // BDA -> BDA -> BDA..
+        bdaService.getBDAs(tdaType1).forEach(tbda -> {
+            DataAttributeRef newDataAttributeRef = DataAttributeRef.copyFrom(dataAttributeRef);
+            newDataAttributeRef.getDaName().getStructNames().add(tbda.getName());
+
+            // STRUCT type (BType=STRUCT) refer to complex BDA object, otherwise it is kind of DA object
+            if(tbda.isSetType() && tbda.getBType().equals(TPredefinedBasicTypeEnum.STRUCT)){
+                daTypeService.findDaType(dtt, tdaType -> tdaType.getId().equals(tbda.getType()))
+                        .ifPresent(nextDaType -> result.addAll(getDataAttributesFromBDA(dtt, nextDaType, newDataAttributeRef)));
+            } else {
+                updateFromAbstractDataAttribute(tbda, newDataAttributeRef.getDaName());
+                result.add(newDataAttributeRef);
+            }
+        });
+        return result;
+    }
+
+    private <T extends TAbstractDataAttribute> void updateFromAbstractDataAttribute(T daOrBda, DaTypeName daTypeName) {
```

Comment:
updating by reference is not a good practice, you hide the real update;  
a good practice is to return the updated object : 
```java
private DaTypeName updateDaNameFromAbstractDataAttribute(TAbstractDataAttribute daOrBda, DaTypeName daTypeName) {
        if (daOrBda.isSetType()) daTypeName.setType(daOrBda.getType());
        if (daOrBda.isSetBType()) daTypeName.setBType(daOrBda.getBType());
        if (daOrBda.isSetValImport()) daTypeName.setValImport(daOrBda.isValImport());
        if (daOrBda.isSetVal()) daTypeName.addDaiValues(daOrBda.getVal());
        return daTypeName;
    }
```
and
```java
newDataAttributeRef.setDaName(updateDaNameFromAbstractDataAttribute(tbda, daName));
```

Reply:
You mean 
```java
newDataAttributeRef.setDaName(updateDaNameFromAbstractDataAttribute(tbda, newDataAttributeRef.getDaName()));
```